src/routes/ProductsPage/ProductsPage.tsx

```diff
@@ -0,0 +1,118 @@
+import type { FC } from 'react';
+import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
+
+import { ControlPanel } from '@components/ControlPanel/ControlPanel';
+import { Pagination } from '@components/Pagination/Pagination';
+import { ProductsList } from '@components/ProductsList/ProductsList';
+import { BASE_URL } from '@constants/apiUrl';
+import { PRODUCTS_PER_PAGE } from '@constants/pagination';
+import { useDataFetching } from '@hooks/useDataFetching';
+import { useFilterAndSort } from '@hooks/useFilterAndSort';
+import { usePagination } from '@hooks/usePagination';
+import type { SortOptionChangeHandler } from '@interfaces/ControlPanel';
+import { SortOption } from '@interfaces/ControlPanel';
+import type { AddProductToCartHandler, Product, ProductCategoryName } from '@interfaces/Product';
+
+import styles from './ProductsPage.module.css';
+
+interface ProductsPageProps {
+    onAddProductToCart: AddProductToCartHandler;
+}
+
+const ProductsPage: FC<ProductsPageProps> = ({ onAddProductToCart }) => {
+    const [searchQuery, setSearchQuery] = useState('');
+    const [selectedFiltersByCategory, setSelectedFilterByCategory] = useState<ProductCategoryName[]>([]);
+    const [selectedSortOption, setSelectedSortOption] = useState<SortOption>(SortOption.PRICE_HIGH_LOW);
+    const [products, setProducts] = useState<Product[]>([]);
+
+    const searchInputReference = useRef<HTMLInputElement>(null);
+
+    const {
+        data: fetchedProducts,
+        isFetching,
+        errorInfo,
+    } = useDataFetching<Product[]>({ url: BASE_URL, parseData: (data) => data.products });
```

Comment:
YOu have an infinite cycle here and ddos our API service))
![image](https://github.com/snischuk/fe-react-2024/assets/98031768/c4ebbd2c-f496-41da-8817-ca77687a4573)
To find the reasons, you have to debug the page. My bet is `parseDate: (data) => data.products` is always new, and it triggers useEffect again and again

Reply:
axaxa I was supposed to be learning frontend... sorry for ddos 😂